cloudinit/config/cc_runcmd.py

```diff
@@ -25,10 +25,17 @@
 
 MODULE_DESCRIPTION = """\
 Run arbitrary commands at a rc.local like level with output to the
-console. Each item can be either a list or a string. If the item is a
-list, it will be properly quoted. Each item is written to
-``/var/lib/cloud/instance/runcmd`` to be later interpreted using
-``sh``.
+console. Each item can be either a list or a string. The item type affects
+how it is executed:
+
+
+* If the item is a string, it will be interpreted by ``sh``.
+* If the item is a list, the items will be executed as if passed to execve(3)
+  (with the first arg as the command).
+
+
+Each item is written to ``/var/lib/cloud/instance/runcmd`` to be later
+interpreted using ``sh``.
```

Comment:
personally, I kind of view that as implementation detail.  And actually a bad implementation detail.

One reason that it is a bad implementation, and one you don't want to advertise to a user, is that it means you can do something like this:

```yaml
runcmd:
 - "foo=bar"
 -  "echo \"foo is $foo\""
```

i don't remember if the above actually would echo 'foo is bar' or not.  But I do recall there was a juju related issue based on the fact that cloud-init was rendering this all to a shell script and executing it.

It would be better if cloud-init handled the execution of these items in the 'runcmd' without  the state that 'sh' usage is providing. 

This is one of those things that you look back at 10 years later and think "why did I do that".

Reply:
Agreed, I'll drop that.